### PR TITLE
Allow cleanup in memory data

### PR DESCRIPTION
**Pull request by evan361425**
It is required when we need to manually allow specific blocked clients to pass the limiter.

### CODE REVIEW

**Comment by animir** [comment]
@evan361425 Hi, thank you for this PR. I think, it is already there. You could use this [delete](https://github.com/animir/node-rate-limiter-flexible/wiki/API-methods#ratelimiterdeletekey) method

**Comment by evan361425** [comment]
But it needs to specify key for deleting storage's data.

In my case, I have to delete all keys. I need this function after deleting all values in DB.

**Comment by animir** [comment]
@evan361425 I see now, thank you for explanation. 

Could you change a couple of things?
1. `cleanupInmemory` function should be named `deleteInMemory`. So it is in sync with other `delete` function name. M is capital as it is correct camelCase.
2. BlockedKeys function should be also named `delete`.
3. Instead of single function for both delete by key and delete all keys, there should be 2 different functions, so developers don't write code, which deletes all keys by mistake. This means two functions on `RateLimiterStoreAbstract` level, BlockedKeys class is internal and it is fine to call the function with single responsibility from it.
4. Tests should be written to covert new lines. We don't need anything complex, just two tests for every new function to make sure keys really deleted by `key` or all keys from `BlockedKeys`.

Feel free to ask any questions, if something doesn't make sense.

File: lib/RateLimiterStoreAbstract.js [hunk 1]
```diff
@@ -284,6 +284,7 @@ module.exports = class RateLimiterStoreAbstract extends RateLimiterAbstract {
     return new Promise((resolve, reject) => {
       this._delete(rlKey, options)
         .then((res) => {
+          this._inmemoryBlockedKeys.delete(key);
```
Comment (evan361425) [inline lib/RateLimiterStoreAbstract.js:287]: It seems like `delete` function is not deleting in-memory data?

**Review by evan361425 — COMMENTED** [review]
Thanks for accepting the PR. The commit fixed 1, 2 and 3.

Where the test should be placed? `test/RateLimiterStoreAbstract.test.js`?

File: lib/RateLimiterStoreAbstract.js (hunk 1 above)
Comment (animir) [reply lib/RateLimiterStoreAbstract.js:287]: @evan361425 Good catch. It should be `this._inmemoryBlockedKeys.delete(rlKey);` (var is `rlKey` - prepared by specific function for keys)

**Comment by animir** [comment]
@evan361425 `test/RateLimiterStoreAbstract.test.js` would be great to put tests. If it is too hard to create an instance of `RateLimiterStoreAbstract`, you can use any store limiter like `RateLimiterRedis` or `RateLimiterMongo` or any other, but put tests inside `test/RateLimiterStoreAbstract.test.js` anyway. It is not ideal, but our aim is to make sure the new logic is covered by tests. I hope, this helps.

File: test/RateLimiterStoreAbstract.test.js [hunk 2]
```diff
@@ -0,0 +1,122 @@
...
+      rateLimiter.consume('key1', 2),
+      rateLimiter.consume('key2', 2),
+    ])
+      .then(() => {
+        expect(rateLimiter._inmemoryBlockedKeys._keys.key1).not.eq(undefined);
+        expect(rateLimiter._inmemoryBlockedKeys._keys.key2).not.eq(undefined);
+
+        rateLimiter.deleteInMemory();
+        expect(rateLimiter._inmemoryBlockedKeys._keys.key1).eq(undefined);
+        expect(rateLimiter._inmemoryBlockedKeys._keys.key2).eq(undefined);
```
Comment (evan361425) [inline test/RateLimiterStoreAbstract.test.js:90]: Confirm deleted

File: test/RateLimiterStoreAbstract.test.js [hunk 3]
```diff
@@ -0,0 +1,122 @@
...
+      inmemoryBlockDuration: 1,
+      keyPrefix: '',
+    });
+
+    // should start blocking
+    Promise.allSettled([
+      rateLimiter.consume('key1', 2),
+      rateLimiter.consume('key2', 2),
+    ])
+      .then(() => {
```
Comment (evan361425) [inline test/RateLimiterStoreAbstract.test.js:84]: `allSettled` will fire `then` even promises rejected

File: test/RateLimiterStoreAbstract.test.js [hunk 4]
```diff
@@ -0,0 +1,122 @@
...
+      keyPrefix: '',
+    });
+
+    // should start blocking
+    rateLimiter.consume('key', 2).catch(() => {
+      expect(rateLimiter._inmemoryBlockedKeys._keys.key).not.eq(undefined);
+
+      rateLimiter.delete('key').then((isExist) => {
+        expect(rateLimiter._inmemoryBlockedKeys._keys.key).eq(undefined);
+        expect(isExist).eq(true);
```
Comment (evan361425) [inline test/RateLimiterStoreAbstract.test.js:116]: Only confirm in memory data is deleted, since confirming DB's data existence should not test here.

**Review by evan361425 — COMMENTED** [review]
The commit add testing.

File: test/RateLimiterStoreAbstract.test.js (hunk 4 above)
Comment (animir) [reply test/RateLimiterStoreAbstract.test.js:116]: right

File: test/RateLimiterStoreAbstract.test.js [hunk 5]
```diff
@@ -0,0 +1,122 @@
+/* eslint-disable security/detect-object-injection */
+const { describe, it } = require('mocha');
+const { expect } = require('chai');
+const RateLimiterStoreAbstract = require('../lib/RateLimiterStoreAbstract');
+const RateLimiterRes = require('../lib/RateLimiterRes');
+
+class RateLimiterStoreMemory extends RateLimiterStoreAbstract {
```
Comment (animir) [inline test/RateLimiterStoreAbstract.test.js:7]: @evan361425 nice. you're ready to implement new rate limiters after this experience :)

File: lib/RateLimiterStoreAbstract.js [hunk 6]
```diff
@@ -356,4 +364,21 @@ module.exports = class RateLimiterStoreAbstract extends RateLimiterAbstract {
...
+   * @param {string} rlKey
+   * @param {number} points
+   * @param {number} msDuration
+   * @param {boolean} forceExpire
+   * @param {Object} options
+   * @abstract
+   *
+   * @return Promise<Object>
+   */
+  _upsert() {
```
Comment (animir) [inline lib/RateLimiterStoreAbstract.js:381]: 👍

**Review by animir — APPROVED** [review]

**Comment by animir** [comment]
@evan361425 thank you!

**Comment by animir** [comment]
@evan361425 After one more review I renamed `deleteInMemory` to `deleteInMemoryBlockedAll`, so it is the same name as options `inmemoryBlockDuration` and `inmemoryBlockOnConsumed`. It may be confusing, since there is RateLimiterMemory. I also added a suffix `All` so it makes even more sense. I hope, you like this change.

**Comment by evan361425** [comment]
You are an excellent reviewer :)

**Comment by animir** [comment]
@evan361425 Thanks

**Pull request merged**